File: Sources/PalmierPro/Models/TextLayout.swift
import AppKit

/// Natural bounding size of a rendered text clip, shared between the layer
/// controller and clip placement.
enum TextLayout {
    static let shadowPadding: CGFloat = 12
    static let referenceCanvasHeight: CGFloat = 1080

    struct WordFrame: Equatable {
        var text: String
        /// Origin and size within the phrase text box (top-left origin).
        var rect: CGRect
    }

    static func naturalSize(
        content: String,
        style: TextStyle,
        maxWidth: CGFloat,
        canvasHeight: CGFloat
    ) -> CGSize {
        let measured = content.isEmpty ? " " : content
        let canvasScale = canvasHeight / referenceCanvasHeight
        let renderSize = CGFloat(style.fontSize * style.fontScale) * canvasScale
        let str = NSAttributedString(
            string: measured,
            attributes: style.attributes(size: renderSize, includeColor: false)
        )
        let bounding = str.boundingRect(
            with: CGSize(width: maxWidth, height: .greatestFiniteMagnitude),
            options: [.usesLineFragmentOrigin, .usesFontLeading]
        )
        // +4px slack absorbs canvas→preview scale rounding.
        let slack: CGFloat = 4
        let shadowPad = style.shadow.enabled ? shadowPadding * 2 : 0
        return CGSize(
            width: max(1, ceil(bounding.width) + shadowPad + slack),
            height: max(1, ceil(bounding.height) + slack)
        )
    }

    static func wordFrames(
        content: String,
        words: [CaptionWordTiming],
        style: TextStyle,
        boxSize: CGSize,
        canvasHeight: CGFloat
    ) -> [WordFrame] {
        guard !content.isEmpty, !words.isEmpty else { return [] }
        let canvasScale = canvasHeight / referenceCanvasHeight
        let fontSize = CGFloat(style.fontSize * style.fontScale) * canvasScale
        let attr = NSAttributedString(string: content, attributes: style.fillAttributes(size: fontSize))
        let storage = NSTextStorage(attributedString: attr)
        let layout = NSLayoutManager()
        storage.addLayoutManager(layout)
        let container = NSTextContainer(size: boxSize)
        container.lineFragmentPadding = 0
        layout.addTextContainer(container)
        layout.ensureLayout(for: container)

        let ns = content as NSString
        let tokenRanges = tokenRanges(in: ns)
        guard !tokenRanges.isEmpty else { return [] }

        var frames: [WordFrame] = []
        for index in words.indices {
            let range = tokenRanges[min(index, tokenRanges.count - 1)]
            let glyphRange = layout.glyphRange(forCharacterRange: range, actualCharacterRange: nil)
            var rect = layout.boundingRect(forGlyphRange: glyphRange, in: container)
            let lineRect = layout.lineFragmentRect(forGlyphAt: glyphRange.location, effectiveRange: nil)
            rect.origin.y = lineRect.origin.y
            rect.size.height = lineRect.height
            frames.append(WordFrame(text: ns.substring(with: range), rect: rect))
        }
        return frames
    }

    private static func tokenRanges(in ns: NSString) -> [NSRange] {
        let whitespace = CharacterSet.whitespacesAndNewlines
        var ranges: [NSRange] = []
        var loc = 0
        while loc < ns.length {
            while loc < ns.length, isWhitespace(ns.character(at: loc), whitespace) { loc += 1 }
            guard loc < ns.length else { break }
            let start = loc
            while loc < ns.length, !isWhitespace(ns.character(at: loc), whitespace) { loc += 1 }
            ranges.append(NSRange(location: start, length: loc - start))
        }
        return ranges
    }

    private static func isWhitespace(_ unit: unichar, _ set: CharacterSet) -> Bool {
        guard let scalar = Unicode.Scalar(unit) else { return false }
        return set.contains(scalar)
    }
}

File: platforms/windows/src/PalmierPro.Core/Models/TextLayout.cs
namespace PalmierPro.Core.Models;

/// `NSAttributedString.boundingRect(with:options:)`'s word-wrapped bounding-box measurement is a
/// render-time text-layout concern (CoreText on Mac, DirectWrite on Windows) — this project has no
/// text-layout stack, so it's the seam `TextLayout.NaturalSize` takes rather than a hard AppKit
/// dependency. The Windows renderer supplies the implementation.
public interface ITextBoundsMeasurer
{
    /// `text` word-wrapped to `maxWidth` (unbounded height), set in `style` at `renderFontSize` pt;
    /// returns the measured (width, height), mirroring `NSAttributedString.boundingRect`.
    (double Width, double Height) MeasureBoundingRect(string text, TextStyle style, double renderFontSize, double maxWidth);
}

/// Natural bounding size of a rendered text clip, shared between the layer controller and clip
/// placement. Ported from TextLayout.swift; actual glyph measurement goes through
/// <see cref="ITextBoundsMeasurer"/> (see its doc comment) rather than a font/text-layout stack.
public static class TextLayout
{
    public const double ShadowPadding = 12;
    public const double ReferenceCanvasHeight = 1080;

    public static (double Width, double Height) NaturalSize(
        string content,
        TextStyle style,
        double maxWidth,
        double canvasHeight,
        ITextBoundsMeasurer measurer)
    {
        var measured = content.Length == 0 ? " " : content;
        var canvasScale = canvasHeight / ReferenceCanvasHeight;
        var renderSize = style.FontSize * style.FontScale * canvasScale;
        var bounding = measurer.MeasureBoundingRect(measured, style, renderSize, maxWidth);

        // +4px slack absorbs canvas -> preview scale rounding.
        const double slack = 4;
        var shadowPad = style.Shadow.Enabled ? ShadowPadding * 2 : 0;
        var borderPad = style.Border.Enabled ? TextStyle.GlyphBorderPadding(renderSize) * 2 : 0;
        return (
            Math.Max(1, Math.Ceiling(bounding.Width) + shadowPad + borderPad + slack),
            Math.Max(1, Math.Ceiling(bounding.Height) + borderPad + slack)
        );
    }
}
